### PR TITLE
feat: Who Am I route

### LINKED ISSUE #2140: Create the /v1/keys.whoami endpoint
### Preliminary Checks

- [X] I have reviewed https://unkey.com/docs for existing features that would solve my problem

- [X] I have searched for existing feature requests: https://github.com/unkeyed/unkey/issues

- [X] This issue is not a question, general help request, or anything other than a feature request directly related to Unkey. Please ask questions in our Discord community: https://unkey.com/discord.

### Is your feature request related to a problem? Please describe.

Sometimes you need to know the current config and limits of a key, but using the [/v1/keys.getKey](https://www.unkey.com/docs/api-reference/keys/get) endpoint requires a `keyId`, which you don't always have available.

### Describe the solution

Create a new `POST /v1/keys.whoami` endpoint in [/apps/api/src/routes](https://github.com/unkeyed/unkey/tree/main/apps/api/src/routes) as well as a testcase for it. Then register the handler in [/apps/api/src/worker.ts](https://github.com/unkeyed/unkey/blob/main/apps/api/src/worker.ts).

I suggest copying another handler file and modifying, so you can reuse the basic structure of the openapi validation.

The endpoint must receive the following payload in a `POST` request:
```ts
{
  key: string
}
```
and be authenticated via a `rootKey` in the `Authorization` header.
The root key must have the `api.${apiId}.read_key` or `api.*.read_key` permissions.

It must return the following json payload:
```ts
{
    id: string,
    name?: string,
    remaining?: number,
    identity?: {
      id: string,
      externalId: string,
    },
    meta?: unknown,
    createdAt: number,
    enabled: boolean,
    environment?: string,
}
```

### General flow of the handler:

1. parse the key from the request
2. hash it using `sha256`, there's an internal library for it
3. look up the key in the db by querying for its hash, include its api and identity relations
4. do a root key auth check, and include the above mentioned permissions
5. return 404 if not found or deleted
6. return structured json as mentioned above

### Describe alternatives you have considered (if any)

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by harshsbhat**
## What does this PR do?

Creates a new API route called Who AM I for keys.

Fixes #2140

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Chore (refactoring code, technical debt, workflow improvements)
- [ ] Enhancement (small improvements)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How should this be tested?

- Test A
- Test B

### Required

- [x] Filled out the "How to test" section in this PR
- [x] Read [Contributing Guide](./CONTRIBUTING.md)
- [x] Self-reviewed my own code
- [x] Commented on my code in hard-to-understand areas
- [x] Ran `pnpm build`
- [x] Ran `pnpm fmt`
- [x] Checked for warnings, there are none
- [x] Removed all `console.logs`
- [x] Merged the latest changes from main onto my branch with `git pull origin main`
- [x] My changes don't cause any responsiveness issues

### Appreciated

- [ ] If a UI change was made: Added a screen recording or screenshots to this PR
- [ ] Updated the Unkey Docs if changes were necessary

## Summary by CodeRabbit

- **New Features**
	- Introduced a new API route at `/v1/keys.whoAmI` for retrieving key information using a query parameter and a POST route for fetching key details.
	- Both routes support bearer authentication and return detailed key information in JSON format.
	- New documentation available for the `/v1/keys.whoAmI` endpoint, detailing its usage and purpose.

- **Bug Fixes**
	- Enhanced error handling for scenarios where keys are not found or database queries encounter issues.

- **Tests**
	- Added integration tests to validate the behavior of the `/v1/keys.whoAmI` endpoint for non-existent keys and permissions based on user roles.
	- Implemented tests to ensure correct responses are returned for valid key requests and identity associations.

### CODE REVIEW

File: apps/api/src/routes/v1_keys_getWhoAmI.ts [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+  tags: ["keys"],
+  operationId: "whoAmI",
+  method: "get",
+  path: "/v1/keys.whoAmI",
+  security: [{ bearerAuth: [] }],
+  request: {
+    query: z.object({
+      key: z.string().min(1).openapi({
+        description: "The actual key to fetch",
+        example: "actual_key_value_here",
```
Comment (chronark) [inline apps/api/src/routes/v1_keys_getWhoAmI.ts:19]: give a concrete example: `sk_lAfRxPiREveSj2WfKVLozw`

File: apps/api/src/routes/v1_keys_getWhoAmI.ts [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+      }),
+    }),
+  },
```
[the remark this answers is not in the document]
Comment (chronark) [reply apps/api/src/routes/v1_keys_getWhoAmI.ts:22]: yup

File: apps/api/src/routes/v1_keys_getWhoAmI.ts [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+  responses: {
+    200: {
+      description: "The configuration for a single key",
+      content: {
+        "application/json": {
+          schema:  z.object({
+            id: z.string().openapi({
+              description: "The ID of the key",
+              example: "abc123",
```
Comment (chronark) [inline apps/api/src/routes/v1_keys_getWhoAmI.ts:31]: that's not what a keyId looks like, they are always prefixed with `key_`
ie: `key_JUviid35Z1wthJhIKJ9AA`

File: apps/api/src/routes/v1_keys_getWhoAmI.ts [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+            }),
+            remaining: z.number().int().optional().openapi({
+              description: "The remaining number of requests for the key",
+              example: 1000,
+            }),
+            identity: z
+              .object({
+                id: z.string().openapi({
+                  description: "The identity ID associated with the key",
+                  example: "identity123",
```
Comment (chronark) [inline apps/api/src/routes/v1_keys_getWhoAmI.ts:45]: that's not what an identity id looks like, they're always prefixed with `id_`-> `id_Ly623bAsiHXwTfWPEVhEQ`

File: apps/api/src/routes/v1_keys_getWhoAmI.ts [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+                  description: "The external identity ID associated with the key",
+                  example: "ext123",
+                }),
+              })
+              .optional()
+              .openapi({
+                description: "The identity object associated with the key",
+              }),
+            meta: z
+              .object({})
```
Comment (chronark) [inline apps/api/src/routes/v1_keys_getWhoAmI.ts:57]: we can narrow this down to `z.record(z.unknown())`

**Review by chronark — CHANGES_REQUESTED** [review]

**Comment by chronark** [comment]
Please add a tests

**Comment by harshsbhat** [comment]
I will get working on this again tomorrow. Had to do some college work today.

**Comment by harshsbhat** [comment]
I have resolved the requested changes. I will add the tests in sometime.

File: apps/api/src/routes/v1_keys_whoAmI.ts [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+    if (key.workspaceId !== auth.authorizedWorkspaceId) {
+      throw new UnkeyApiError({
+        code: "NOT_FOUND",
+        message: "Key not found",
+      });
+    }
+    let meta = key.meta ? JSON.parse(key.meta) : undefined;
+    if (!meta || Object.keys(meta).length === 0) {
+      meta = undefined;
+    }
```
[the remark this answers is not in the document]
Comment (harshsbhat) [reply apps/api/src/routes/v1_keys_whoAmI.ts:154]: We check the metadata is valid data while creating the key. So there is probably a 0% chance that while getting the key information the meta is not a JSON right? wdyt @chronark

[the remark this answers is not in the document]
Comment (chronark) [reply apps/api/src/routes/v1_keys_whoAmI.ts:154]: @harshsbhat yes, it's guaranteed to be valid

**Comment by harshsbhat** [comment]
@chronark  I am not very confident about the tests as I don't have too much experience writing tests. But can you review these two? I have referred to the other tests to write these tests. If everything is good I can write the third one too. That way we can distribute the load of reviews between these two.

File: apps/api/src/routes/v1_keys_whoAmI.security.test.ts [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    const hash = await sha256(key);
+    await h.db.primary.insert(schema.keys).values({
+      id: keyId,
+      keyAuthId: h.resources.userKeyAuth.id,
+      hash: hash,
+      start: key.slice(0, 8),
+      workspaceId: h.resources.userWorkspace.id,
+      createdAt: new Date(),
+    });
+    return {
```
[the remark this answers is not in the document]
Comment (chronark) [reply apps/api/src/routes/v1_keys_whoAmI.security.test.ts:25]: @harshsbhat the testharness has a utiility for this already `h.createKey` I believe

**Comment by chronark** [comment]
@harshsbhat yeah those look fine, I just made one comment.


For the happy tests, just do 2 cases, one where there is an identity and one without and then assert that the response is as expected

**Comment by harshsbhat** [comment]
@chronark should be good for review

File: apps/docs/api-reference/keys/whoami.mdx [hunk 8]
```diff
@@ -0,0 +1,15 @@
+---
+title: Returns data about a key
+openapi: get /v1/keys.whoami
+---
+
+## Changelog
+
+| Date        | Changes                     |
+|-------------|-----------------------------|
+| Oct 07 2024 | Introduced endpoint         |
```
[the remark this answers is not in the document]
Comment (chronark) [reply apps/docs/api-reference/keys/whoami.mdx:10]: what makes you think it's not the correct date?

**Comment by harshsbhat** [comment]
Enjoyed every minute of writing this new route. Thanks @chronark  for giving an opportunity for developers like me ( with no experience ) to actually see my code in the production. I would love to update the docs for this too

**Pull request merged**